src/main/native/pdhmm/pdhmm.h

```diff
@@ -261,16 +293,31 @@ inline VEC_INT_TYPE toPrime(VEC_INT_TYPE x)
     return result;
 }
 
-inline void initializePriorsVec(const INT_TYPE *haplotypeBases, const INT_TYPE *haplotypePDBases, const INT_TYPE *readBases, INT_TYPE *readQuals, INT_TYPE startIndex, INT_TYPE max_hap_bases_length, INT_TYPE max_read_bases_length, double *priorVec, const double *qualToErrorProbCache, int32_t maxHaplotypeLength)
+inline int32_t initializePriorsVec(const INT_TYPE *haplotypeBases, const INT_TYPE *haplotypePDBases, const INT_TYPE *readBases, INT_TYPE *readQuals, INT_TYPE startIndex, INT_TYPE currMaxHaplotypeLength, INT_TYPE currMaxReadLength, double *priorVec, const double *qualToErrorProbCache, int32_t maxHaplotypeLength, INT_TYPE readArrayLen, INT_TYPE hapArrayLen, INT_TYPE priorVecLen)
 {
     // initialize the prior matrix for all combinations of read x haplotype bases
     INT_TYPE paddedMaxHaplotypeLength = maxHaplotypeLength + 1;
     VEC_INT_TYPE nVec = VEC_SET1_INT((int8_t)'N');
     VEC_INT_TYPE aVec = VEC_SET1_INT('a');
     VEC_INT_TYPE caseDiff = VEC_SET1_INT(32);
 
+    if (currMaxReadLength * SIMD_WIDTH_DOUBLE > readArrayLen)
+    {
+        return PDHMM_MEMORY_ACCESS_ERROR;
+    }
+
+    if (currMaxHaplotypeLength * SIMD_WIDTH_DOUBLE > hapArrayLen)
+    {
+        return PDHMM_MEMORY_ACCESS_ERROR;
+    }
+
+    if ((currMaxReadLength + 1) * paddedMaxHaplotypeLength * SIMD_WIDTH_DOUBLE > priorVecLen)
+    {
+        return PDHMM_MEMORY_ACCESS_ERROR;
+    }
+
     // manual unroll rows
-    INT_TYPE n = (max_read_bases_length / 4) * 4;
+    INT_TYPE n = (currMaxReadLength / 4) * 4;
```

Comment:
In C/C++, when you divide an integer, it automatically rounds down to the nearest whole number. So, dividing and then multiplying an integer by the same number gives you the closest multiple of that number which is less than or equal to the original integer. 

Here, `n` is set to the nearest multiple of 4 that is less than or equal to `currMaxReadLength `.

Reply:
This would work not only in C/C++ but in any language with integer division but it looks like a mistake at first glance. I would recommend to create a helper function with clear name, eg:

`inline int roundDownToNearestMultipleOf(int val, int mul) {
   return (val/mul)*mul;
}`